src/python/pants/backend/python/lint/pydocstyle/subsystem.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+from pants.backend.python.goals import lockfile
+from pants.backend.python.goals.export import ExportPythonTool, ExportPythonToolSentinel
+from pants.backend.python.goals.lockfile import (
+    GeneratePythonLockfile,
+    GeneratePythonToolLockfileSentinel,
+)
+from pants.backend.python.lint.pydocstyle.skip_field import SkipPydocstyleField
+from pants.backend.python.subsystems.python_tool_base import ExportToolOption, PythonToolBase
+from pants.backend.python.subsystems.setup import PythonSetup
+from pants.backend.python.target_types import (
+    ConsoleScript,
+    InterpreterConstraintsField,
+    PythonSourceField,
+)
+from pants.backend.python.util_rules.partition import _find_all_unique_interpreter_constraints
+from pants.core.goals.generate_lockfiles import GenerateToolLockfileSentinel
+from pants.core.util_rules.config_files import ConfigFilesRequest
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.option.option_types import ArgsListOption, FileOption, SkipOption
+from pants.util.docutil import git_url
+from pants.util.logging import LogLevel
+from pants.util.strutil import softwrap
+
+
+@dataclass(frozen=True)
+class PydocstyleFieldSet(FieldSet):
+    required_fields = (PythonSourceField,)
+
+    source: PythonSourceField
+    interpreter_constraints: InterpreterConstraintsField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipPydocstyleField).value
+
+
+class Pydocstyle(PythonToolBase):
+    options_scope = "pydocstyle"
+    name = "Pydocstyle"
+    help = "A tool for checking compliance with Python docstring conventions (http://www.pydocstyle.org/en/stable/)."
+
+    default_version = "pydocstyle>=6.1.1,<7.0"
+    default_main = ConsoleScript("pydocstyle")
+
+    register_lockfile = True
+    default_lockfile_resource = ("pants.backend.python.lint.pydocstyle", "pydocstyle.lock")
+    default_lockfile_path = "src/python/pants/backend/python/lint/pydocstyle/pydocstyle.lock"
+    default_lockfile_url = git_url(default_lockfile_path)
+
+    skip = SkipOption("lint")
+    args = ArgsListOption(example="--select=D101,D102")
+    export = ExportToolOption()
+    config = FileOption(
+        default=None,
+        advanced=True,
+        help="Path to a Pydocstyle config file (http://www.pydocstyle.org/en/stable/usage.html#configuration-files).",
+    )
```

Comment:
I think we also want to use config autodiscovery:

```
In order for pydocstyle to use a configuration file automatically, it must be named one of the following options.

setup.cfg
tox.ini
.pydocstyle
.pydocstyle.ini
.pydocstylerc
.pydocstylerc.ini
pyproject.toml
```

https://github.com/pantsbuild/pants/blob/2fb6d6f204d43b62fa2d38f9b7b2b34ae307284a/src/python/pants/backend/python/lint/flake8/subsystem.py#L101-L113

https://github.com/pantsbuild/pants/blob/2fb6d6f204d43b62fa2d38f9b7b2b34ae307284a/src/python/pants/backend/python/lint/flake8/subsystem.py#L152-L162

I'm unclear if those files must be at the root of the repository, vs can be in directories higher up. If you're willing, it would be helpful to experiment to see what the behavior is.

Reply:
Thanks for suggesting this, that's super helpful. Yes, this is what the docs say:

> When searching for a configuration file, pydocstyle looks for one of the file specified above in that exact order. ini-like configuration files must have a [pydocstyle] section while toml configuration files must have a [tool.pydocstyle] section. If a configuration file was not found, pydocstyle keeps looking for one up the directory tree until one is found or uses the default configuration.

Reply:
http://www.pydocstyle.org/en/stable/usage.html#configuration-files

I am not sure if we can/want to traverse the directory up as the `./pants` command will be run from the root repo? So the config files are likely to be defined in the root.

Reply:
I think they paragraph is ambiguous if it starts looking in CWD or relative to the file argument(s). If the former, then yeah we can only look in the build root. If the latter we have code for that

Reply:
> I think they paragraph is ambiguous

@AlexTereshenkov It would be helpful to experiment with this to figure out what the behavior is

Reply:
I have experimented and can confirm that when running `pydocstyle helloworld/main.py` with the config file located in the parent directory, it does actually walks up the directory tree (outside the repo root)! I find this behavior very surprising as I can't think why I would want my config file to be picked up from `/home/user.name` when I run `pydocstyle` in the `/home/user.name/code/repo`. I'd very much like to avoid having any outside of the repo config files leak into the repo I am working with.

@thejcannon what do you say if we document that Pants will only search for the config files in the repository root (from where the `./pants` command ran) OR by using the config provided in the `[pydocstyle].config`? I think this is what we do for all other formatters/linters, right?

Reply:
Yeah, feel free to elaborate in the docs. FWIW I also expect users to share your expectation.

Reply:
Yes, that's a great idea. I've extended the help.